# Tugas Praktikum Data Mining TM-9
## Nama: Faiz Iqbal I'tishom
## NIM: 164231059
## Kelas: SD-A1

### Import Library

In [172]:
import numpy as np

### Inisialisasi input, trageted value, bobot, dan bias pada fungsi


In [173]:
X = np.array([1.0, 2.0, -1.0])
y_true = 3.5
W1 = np.random.randn(4, 3)
b1 = np.random.randn(4)
W2 = np.random.randn(1, 4)
b2 = np.random.randn(1)

### Inisialisasi fungsi aktivasi

In [174]:
def relu(x):
    return np.maximum(0, x)

def mean_squared_error(y_pred, y_true):
    return 0.5 * (y_pred - y_true) ** 2

### Forward Pass

In [175]:
def forward_pass(X, W1, b1, W2, b2):
    hidden_input = np.dot(W1, X) + b1
    hidden_output = relu(hidden_input)
    final_input = np.dot(W2, hidden_output) + b2
    y_pred = final_input[0]
    return y_pred, hidden_output, hidden_input, final_input

### Backward Pass

In [176]:
def backward_pass(X, y_true, W1, b1, W2, b2, hidden_output, hidden_input, output_input, y_pred):
    dL_dy_pred = y_pred - y_true
    dW2 = dL_dy_pred * hidden_output.reshape(1, -1)
    db2 = dL_dy_pred

    dL_da1 = dL_dy_pred * W2.reshape(-1)
    dL_dz1 = dL_da1 * relu_derivative(hidden_input)

    dW1 = dL_dz1.reshape(-1, 1) @ X.reshape(1, -1)
    db1 = dL_dz1
    return dW2, db2, dW1, db1

## Backpropagation

### Inisialisasi seed, learning rate, dan epoch

In [177]:
# Seed agar reproducible
np.random.seed(27229)
# Set format tampilan array agar lebih rapi
np.set_printoptions(precision=4, suppress=True)
# Input dan targeted value
learning_rate = 0.1
epochs = 3

### Inisialisasi Fungsi Relu dan Turunannya

In [178]:
def relu(x):
    return np.maximum(0, x)
def relu_derivative(x):
    return (x > 0).astype(float)

### Forward Pass dan Backpropagation

In [179]:
def train_regression_3epoch():
    np.random.seed(27229)
    X = np.array([1.0, 2.0, -1.0])
    y_true = 3.5
    W1 = np.random.randn(4, 3)
    b1 = np.random.randn(4)
    W2 = np.random.randn(1, 4)
    b2 = np.random.randn(1)
    learning_rate = 0.1

    for epoch in range(epochs):
        print(f"\n{'='*25}\nEpoch {epoch + 1}\n{'='*25}")

        # Forward pass
        y_pred, hidden_output, hidden_input, final_input = forward_pass(X, W1, b1, W2, b2)
        loss = 0.5 * (y_true - y_pred) ** 2

        print("\n>> INPUT")
        print(f"X          : {X}")
        print(f"Target y   : {y_true:.4f}")

        print("\n>> FORWARD PASS")
        print(f"W1         :\n{W1}")
        print("\n")
        print(f"b1         : {b1}")
        print("\n")
        print(f"z1 (Hidden Input)  : {hidden_input}")
        print("\n")
        print(f"a1 (Hidden Output) : {hidden_output}")
        print("\n")
        print(f"W2         :\n{W2}")
        print("\n")
        print(f"b2         : {b2}")
        print("\n")
        print(f"z2 (Final Input)   : {final_input}")
        print("\n")
        print(f"y_pred     : {y_pred:.4f}")
        print("\n")
        print(f"Loss       : {loss:.6f}")

        # Backward pass
        dW2, db2, dW1, db1 = backward_pass(X, y_true, W1, b1, W2, b2, hidden_output, hidden_input, final_input, y_pred)

        print("\n>> BACKWARD PASS (Gradients)")
        print(f"dW2        :\n{dW2}")
        print("\n")
        print(f"db2        : {db2:.6f}")
        print("\n")
        print(f"dW1        :\n{dW1}")
        print("\n")
        print(f"db1        : {db1}")

        # Update
        W2 -= learning_rate * dW2
        b2 -= learning_rate * db2
        W1 -= learning_rate * dW1
        b1 -= learning_rate * db1

        print("\n>> UPDATED PARAMETERS")
        print(f"Updated W1 :\n{W1}")
        print("\n")
        print(f"Updated b1 : {b1}")
        print("\n")
        print(f"Updated W2 :\n{W2}")
        print("\n")
        print(f"Updated b2 : {b2}")
        print("\n")
        print(f"y_pred : {y_pred:.4f} | Loss : {loss:.6f} | Akurasi: {1 - abs(y_pred - y_true):.6f} | ")

# Jalankan training
train_regression_3epoch()


Epoch 1

>> INPUT
X          : [ 1.  2. -1.]
Target y   : 3.5000

>> FORWARD PASS
W1         :
[[-1.3719 -0.7152 -2.2667]
 [ 1.0523  0.6514  1.014 ]
 [-0.5866 -0.2038  0.907 ]
 [ 0.7163 -0.1347  0.841 ]]


b1         : [-1.313   0.3261  0.8935  0.0701]


z1 (Hidden Input)  : [-1.8486  1.6671 -1.0077 -0.3241]


a1 (Hidden Output) : [0.     1.6671 0.     0.    ]


W2         :
[[-1.1149  0.8154  0.6199  0.6267]]


b2         : [0.4025]


z2 (Final Input)   : [1.7619]


y_pred     : 1.7619


Loss       : 1.510436

>> BACKWARD PASS (Gradients)
dW2        :
[[-0.     -2.8975 -0.     -0.    ]]


db2        : -1.738066


dW1        :
[[ 0.      0.      0.    ]
 [-1.4172 -2.8345  1.4172]
 [ 0.      0.      0.    ]
 [ 0.      0.      0.    ]]


db1        : [ 0.     -1.4172 -0.     -0.    ]

>> UPDATED PARAMETERS
Updated W1 :
[[-1.3719 -0.7152 -2.2667]
 [ 1.194   0.9348  0.8723]
 [-0.5866 -0.2038  0.907 ]
 [ 0.7163 -0.1347  0.841 ]]


Updated b1 : [-1.313   0.4678  0.8935  0.0701]


Updated W2

## Interpretasi

### Epoch 1

#### Forward Pass
Hidden input (z1): Dihasilkan dari W1 * X + b1

Hidden output (a1): Hanya neuron ke-2 aktif (ReLU > 0)

Final input (z2) dan y_pred: 1.7619 → masih jauh dari target 3.5

#### Backward Pass
Loss: 1.5104 → cukup besar

Gradien besar di W2 dan W1, khususnya neuron ke-2 → wajar karena hanya neuron itu yang aktif

Parameter update: Terjadi perubahan signifikan pada bobot dan bias neuron ke-2

#### Akurasi: -0.738 → Negatif karena error masih besar

### Epoch 2

#### Forward Pass
Hidden input dan output meningkat karena bobot neuron ke-2 sudah diperbaiki

Prediksi (y_pred) naik drastis ke 3.5152, sangat dekat ke target 3.5

Loss turun drastis ke 0.000115

#### Backward Pass
Gradien sangat kecil, karena prediksi hampir tepat

Bobot dan bias hanya mengalami penyesuaian kecil → ini mencerminkan konvergensi

#### Akurasi: 0.9848 → Hampir sempurna

### Epoch 3

#### Forward Pass
y_pred semakin dekat ke 3.5: 3.4900

Loss makin kecil: 0.000050

Gradien makin kecil → menandakan bahwa model sudah mencapai titik optimal lokal

#### Backward Pass
Pembaruan bobot hampir tidak signifikan → training hampir selesai

#### Akurasi: 0.9900 → Sangat tinggi

### Kesimpulan

#### Model belajar sangat efektif hanya dalam 3 epoch karena input sederhana dan jaringan kecil.

#### Neuron kedua hidden layer adalah yang paling aktif dan berperan penting dalam output akhir.

#### Proses ini menunjukkan konvergensi cepat, cocok untuk data dan target sederhana.